### PR TITLE
Fix TaskManagerTest

### PR DESCRIPTION
**Pull request by vvcephei**
Finishing up the TaskManagerTests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
@guozhangwang , I have a minute, so I figured I could fix this. I fixed the failing tests, but there are still ignored tests that need to be re-enabled.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 1]
```diff
@@ -356,27 +347,14 @@ public void shouldCloseStandbyTasksOnShutdown() {
         verify(activeTaskCreator);
     }
 
-    @Ignore
-    @Test
-    public void shouldUnassignChangelogPartitionsOnShutdown() {
-        throw new RuntimeException();
-//        restoreConsumer.unsubscribe();
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:363]: Added this check to shouldCloseActiveTasksOnShutdown (although now, it's a changelogReader expectation)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 2]
```diff
@@ -710,64 +688,22 @@ public void shouldReturnFalseWhenThereAreStillNonRunningTasks() {
...
-//        EasyMock.replay(active, consumer, changeLogReader);
-//
-//        // shouldn't invoke `resume` method in consumer
-//        taskManager.updateNewAndRestoringTasks();
-//        verify(consumer);
-    }
-
-    @Ignore
-    @Test
-    public void shouldUpdateTasksFromPartitionAssignment() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:737]: This is no longer relevant.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 3]
```diff
@@ -710,64 +688,22 @@ public void shouldReturnFalseWhenThereAreStillNonRunningTasks() {
...
         assertThat(taskManager.checkForCompletedRestoration(), is(false));
         assertThat(task00.state(), is(Task.State.RESTORING));
+        // this could be a bit mysterious; we're verifying _no_ interactions on the consumer,
+        // since the taskManager should _not_ resume the assignment while we're still in RESTORING
+        verify(consumer);
     }
 
-    @Ignore
-    @Test
-    public void shouldNotResumeConsumptionUntilAllStoresRestored() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:722]: This is tested effectively by making consumer a strict mock.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -127,39 +128,21 @@ public void setUp() {
...
+        replay(activeTaskCreator, topologyBuilder);
 
-        EasyMock.replay(activeTaskCreator,
-                        standbyTaskCreator,
-                        topologyBuilder,
-                        subscriptionUpdates);
-    }
-
-    @Test
-    public void shouldNotUpdateSubscriptionFromAssignmentOfExistingTopic1() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:149]: If I read it right, this test is unnecessary now that we just pass through the whole assignment and let the topology builder itself merge the new subscription with the old one. I.e., this test is the same as the prior one now.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 5]
```diff
@@ -356,133 +351,57 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
+        assertThat(taskManager.checkForCompletedRestoration(), is(true));
 
         assertThat(task01.state(), is(Task.State.RUNNING));
         assertThat(taskManager.activeTaskMap(), Matchers.anEmptyMap());
         assertThat(taskManager.standbyTaskMap(), Matchers.equalTo(singletonMap(taskId01, task01)));
     }
 
-    @Ignore
-    @Test
-    public void shouldResumeRestoredPartitions() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:409]: Does this need to be a ChangelogReader test now?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 6]
```diff
@@ -356,133 +351,57 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
-//        expectLastCall();
-//        replay();
-//
-//        taskManager.updateNewAndRestoringTasks();
-//        verify(consumer);
-    }
-
-    @Ignore
-    @Test
-    public void shouldReturnTrueWhenActiveAndStandbyTasksAreRunning() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:425]: This is asserted all over the other tests now

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 7]
```diff
@@ -356,133 +351,57 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
-//        mockAssignStandbyPartitions(1L);
-//        expect(standby.allTasksRunning()).andReturn(true);
-//        replay();
-//
-//        assertTrue(taskManager.updateNewAndRestoringTasks());
-    }
-
-    @Ignore
-    @Test
-    public void shouldReturnFalseWhenOnlyActiveTasksAreRunning() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:436]: I honestly wasn't sure what this test was going for... maybe it meant when only _standby_ tasks are running?

Even so, that hardly seems like an important behavior to verify. Just to be clear, behavior of the manager is to return true when all its tasks are in RUNNING state, regardless of the task type.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 8]
```diff
@@ -356,133 +351,57 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
-//        mockAssignStandbyPartitions(1L);
-//        expect(standby.allTasksRunning()).andReturn(false);
-//        replay();
-//
-//        assertFalse(taskManager.updateNewAndRestoringTasks());
-    }
-
-    @Ignore
-    @Test
-    public void shouldSeekToCheckpointedOffsetOnStandbyPartitionsWhenOffsetGreaterThanEqualTo0() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:447]: should this be a ChangelogReader test now?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 9]
```diff
@@ -356,133 +351,57 @@ public void shouldCloseStandbyTasksOnShutdown() {
...
-//        expectLastCall();
-//        replay();
-
-//        taskManager.updateNewAndRestoringTasks();
-//        verify(restoreConsumer);
-    }
-
-    @Ignore
-    @Test
-    public void shouldSeekToBeginningIfOffsetIsLessThan0() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:460]: should this be a ChangelogReader test now?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 10]
```diff
@@ -667,36 +652,21 @@ public boolean maybePunctuateSystemTime() {
...
-        taskManager.checkForCompletedRestoration();
+        assertThat(taskManager.checkForCompletedRestoration(), is(true));
 
         assertThat(task00.state(), is(Task.State.RUNNING));
 
         // one for stream and one for system time
         assertThat(taskManager.punctuate(), equalTo(2));
     }
 
-    // TODO K9113: the following three tests needs to be fixed once thread calling restore is cleaned
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:685]: This test still needs to be added elsewhere, too.

**Review by vvcephei — COMMENTED** [review]
Hey @guozhangwang ,

I managed to finish up the failing/ignored tests in TaskManagerTest this evening. There are several tests of logic that has been moved elsewhere in Streams, so I made a note of them on this PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:149]: Sounds good.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:409]: Yes, it should be, I can add one in ChangelogReaderTest.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:436]: Yup, I agree.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:447]: Yes, this is already in the test class.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:460]: Yes, already in StoreChangelogReaderTest.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:685]: Ack, this is covered in the ChangelogReaderTest now.

**Review by guozhangwang — COMMENTED** [review]
Thanks @vvcephei ! I will merge the PR now and try to finish up the rest of the tests.

At the moment I'm adding more coverage to StreamTaskTest after I've completed the suspend / commit / close logic branching on state(), if you can take a quick look and lmk what do you think on these three functions that would be great :)

**Comment by vvcephei** [comment]
Thanks @guozhangwang ! I just took a look at those methods in `base`. I think they look the same as the last time i saw them, but did confirm they are missing coverage. I'll be available sporadically for reviews, if you want to send a PR for your tests.

**Pull request merged**